studiopress/pattern-manager
# Try using tests/lints/builds inside Docker Containers

## Tasks

- [x] I have signed the [Contributor License Agreement](https://wpeng.in/cla/) with WP Engine and understand that this plugin is not accepting community PRs for new features or major bug fixes with risk of regression. The PR I'm submitting is a minor bug fix.

### Summary of changes
The main change here is that everything runs in docker containers, so you don't even need node installed on your machine. That means that instead of baking everything into package.json (which requires node), we just use a bash script in the plugin root. 

This means that you can run linting/testing without needing to have node, php, composer, or anything except Docker (Docker Desktop on Mac/Windows).

Note that it's also cloned inside the plugin now in a hidden folder called `.pluginade`, and the version is pinned to use a specific tag of pluginade-scripts inside the pluginade.sh file. 

### Notes: 
#### WPCS has been updated to latest:
https://github.com/pluginade/pluginade-scripts/blob/318965c971e92a355c6bc6d1b4e0040de4c643bd/composer.json#L17

That meant a few unimportant changes were causing phpcs to fail, so a custom phpcs.xml has been added to Pattern Manager, so we can define our own custom standards.

I have also fixed a few things that the linter is now catching that I felt we could comply with. 

#### PHPUnit
Additionally, because PHPUnit is now running in a docker container that isn't wp-env, the default for the environment type of `production`, and some of our environment checking functions were difficult to test outside of that, so I have changed them to accept a paramater for the type of environment so we can test it more easily. I modified those tests and the usage of `check_environment_notice_should_show`.

#### The way we run testing and linting will slightly change:

In order to run pluginade commands, actions would be changed as follows:

- `npm run dev` becomes `sh pluginade.sh dev`
- `npm run build` becomes `sh pluginade.sh build`
- `npm run lint:php` becomes `sh pluginade.sh lint:php`
- `npm run lint:php:fix` becomes `sh pluginade.sh lint:php:fix`
- `npm run lint:js` becomes `sh pluginade.sh lint:js`
- `npm run lint:js:fix` becomes `sh pluginade.sh lint:js:fix`
- `npm run lint:css` becomes `sh pluginade.sh lint:css`
- `npm run lint:css:fix` becomes `sh pluginade.sh lint:css:fix`
- `npm run lint:css:fix` becomes `sh pluginade.sh lint:css:fix`

### How to test
1. Notice the circleCi tests, and how they now operate.
2. Try running the above commands on your local machine.
